### PR TITLE
Scripts optparse to argparse

### PR DESCRIPTION
**Pull request by stevepiercy**
refs: https://github.com/Pylons/pyramid/pull/2842

### CODE REVIEW

File: pyramid/scripts/prequest.py [hunk 1]
```diff
@@ -139,10 +156,10 @@ def run(self):
                 name, value = item.split(':', 1)
                 headers[name] = value.strip()
 
-        app = self.get_app(app_spec, self.options.app_name,
-                options=parse_vars(self.args[2:]))
+        app = self.get_app(app_spec, self.args.app_name,
+                options=self.args.config_vars)
```
Comment (mmerickel) [inline pyramid/scripts/prequest.py]: missing `parse_vars`

File: pyramid/scripts/pviews.py [hunk 2]
```diff
@@ -28,16 +27,21 @@ class PViewsCommand(object):
     """
     stdout = sys.stdout
 
-    parser = optparse.OptionParser(
-        usage,
+    parser = argparse.ArgumentParser(
+        prog="pviews",
```
Comment (mmerickel) [inline pyramid/scripts/pviews.py]: drop `prog` argument

File: pyramid/tests/test_scripts/test_proutes.py [hunk 3]
```diff
@@ -38,7 +38,8 @@ def _makeConfig(self, *arg, **kw):
     def test_good_args(self):
         cmd = self._getTargetClass()([])
         cmd.bootstrap = (dummy.DummyBootstrap(),)
-        cmd.args = ('/foo/bar/myapp.ini#myapp', 'a=1')
+        cmd.args.config_uri = ('/foo/bar/myapp.ini#myapp',)
```
Comment (mmerickel) [inline pyramid/tests/test_scripts/test_proutes.py]: not a tuple

File: pyramid/tests/test_scripts/test_pserve.py [hunk 4]
```diff
@@ -36,10 +36,10 @@ def test_run_no_args(self):
         self.assertEqual(result, 2)
         self.assertEqual(self.out_.getvalue(), 'You must give a config file')
 
-    def test_get_options_no_command(self):
+    def test_config_args_no_command(self):
         inst = self._makeOne()
-        inst.args = ['foo', 'a=1', 'b=2']
-        result = inst.get_options()
+        inst.args.config_args = ['a=1', 'b=2']
```
Comment (mmerickel) [inline pyramid/tests/test_scripts/test_pserve.py]: `config_vars`

**Comment by stevepiercy** [comment]
ptweens and pviews are ready for review.

I'm stuck on pshell, and I don't know where to go next. I think I got pshell to run as follows. Using my local project from the starter scaffold, I did `pip install -e .` to pull in all the latest changes of the p* scripts, then tried to run `../../env35/bin/pshell development.ini`. This started an interactive Python shell.

There are probably mistakes in my pshell, too, the latest of which is committed and pushed here.

In the tests (latest also committed and pushed), there's `patch_*` stuff in `_makeOne()`. I tried debugging the test, but I could not figure out where to look or what it does, so I don't know how to fix the tests.

**Comment by mmerickel** [comment]
You need to fix the `_makeOne` method to do the right thing. It doesn't match up with all of the changes you're making.

**Comment by stevepiercy** [comment]
All things pass, except docs continue to fail to build (due to Sphinx/docutils bug), but build locally. Hopefully Sphinx will make a new release with the fix.
https://travis-ci.org/Pylons/pyramid/jobs/183174559#L700

Ready for final review. Can you believe it?

File: pyramid/scripts/pviews.py [hunk 5]
```diff
@@ -230,16 +247,16 @@ def output_view_info(self, view_wrapper, level=1):
                 self.out("%sview predicates (%s)" % (indent, predicate_text))
 
     def run(self):
-        if len(self.args) < 2:
+        if not self.args.config_uri and not self.args.url:
```
Comment (mmerickel) [inline pyramid/scripts/pviews.py]: **or** not **and**

File: pyramid/scripts/prequest.py [hunk 6]
```diff
@@ -106,11 +123,11 @@ def configure_logging(self, app_spec):
         setup_logging(app_spec)
 
     def run(self):
-        if not len(self.args) >= 2:
+        if not self.args.config_uri and not self.args.path_info:
```
Comment (mmerickel) [inline pyramid/scripts/prequest.py]: **or** not **and**

**Comment by mmerickel** [comment]
I tested the scripts by hand and things seem to basically work aside from the two comments above.

The output for ``pcreate --help`` is pretty redundant right now. It happily emits the deprecation warning twice. Once as part of help and once as part of running the command. It would be nice to move the warning lower down in the logic to avoid it being emitted for invalid commands and help.

File: pyramid/scripts/pviews.py (hunk 5 above)
Comment (stevepiercy) [inline pyramid/scripts/pviews.py]: Fixed in 258f8482549a75a627e5b80790ca7f7f84308684

File: pyramid/scripts/prequest.py (hunk 6 above)
Comment (stevepiercy) [inline pyramid/scripts/prequest.py]: Fixed in 258f8482549a75a627e5b80790ca7f7f84308684

**Comment by bertjwregeer** [comment]
I'll review this in a little bit.

File: docs/conf.py [hunk 7]
```diff
@@ -51,11 +51,11 @@ def nothing(*arg):
 extensions = [
     'sphinx.ext.autodoc',
     'sphinx.ext.doctest',
+    'repoze.sphinx.autointerface',
```
Comment (bertjwregeer) [inline docs/conf.py]: Is there a particular reasoning for the ordering of these? Or can they be alphabetized?

**Comment by mmerickel** [comment]
Thanks for powering through @stevepiercy

**Comment by stevepiercy** [comment]
Thank *you* for your guidance and patience. Hope the multiple facepalm wounds heal.

**Comment by stevepiercy** [comment]
Closes #2804

**Pull request merged**